src/Functions/FunctionsConversion.h

```diff
@@ -1993,6 +1993,124 @@ struct ConvertImplGenericFromString
     }
 };
 
+struct ConvertImplFromDynamicToColumn
+{
+    static ColumnPtr execute(
+        const ColumnsWithTypeAndName & arguments,
+        const DataTypePtr & result_type,
+        size_t input_rows_count,
+        const std::function<ColumnPtr(ColumnsWithTypeAndName &, const DataTypePtr)> & nested_convert)
+    {
+        /// When casting Dynamic to regular column we should cast all variants from current Dynamic column
+        /// and construct the result based on discriminators.
+        const auto & column_dynamic = assert_cast<const ColumnDynamic &>(*arguments.front().column.get());
+        const auto & variant_column = column_dynamic.getVariantColumn();
+        const auto & variant_info = column_dynamic.getVariantInfo();
+
+        /// First, cast usual variants to result type.
+        const auto & variant_types = assert_cast<const DataTypeVariant &>(*variant_info.variant_type).getVariants();
+        std::vector<ColumnPtr> cast_variant_columns;
+        cast_variant_columns.reserve(variant_types.size());
+        for (size_t i = 0; i != variant_types.size(); ++i)
+        {
+            /// Skip shared variant, it will be processed later.
+            if (i == column_dynamic.getSharedVariantDiscriminator())
+            {
+                cast_variant_columns.push_back(nullptr);
+                continue;
+            }
+
+            ColumnsWithTypeAndName new_args = arguments;
+            new_args[0] = {variant_column.getVariantPtrByGlobalDiscriminator(i), variant_types[i], ""};
+            cast_variant_columns.push_back(nested_convert(new_args, result_type));
+        }
+
+        /// Second, collect all variants stored in shared variant and cast them to result type.
+        std::vector<MutableColumnPtr> variant_columns_from_shared_variant;
+        DataTypes variant_types_from_shared_variant;
+        /// We will need to know what variant to use when we see discriminator of a shared variant.
+        /// To do it, we remember what variant was extracted from each row and what was it's offset.
+        PaddedPODArray<UInt64> shared_variant_indexes;
+        PaddedPODArray<UInt64> shared_variant_offsets;
+        std::unordered_map<String, UInt64> shared_variant_to_index;
+        const auto & shared_variant = column_dynamic.getSharedVariant();
+        const auto shared_variant_discr = column_dynamic.getSharedVariantDiscriminator();
+        const auto & local_discriminators = variant_column.getLocalDiscriminators();
+        const auto & offsets = variant_column.getOffsets();
+        if (!shared_variant.empty())
+        {
+            shared_variant_indexes.reserve(input_rows_count);
+            shared_variant_offsets.reserve(input_rows_count);
+            FormatSettings format_settings;
```

Comment:
I see other places use also a default constructed FormatSettings instead of one based on context settings. Is that the way to go when calling `deserializeBinary`?

Reply:
Yes, because we don't want any settings to change the internal representation of the value inside SharedVariant. For example, there is a setting `output_format_binary_write_json_as_string` that can change the way how we serialize values of nested JSON, and it can break the deserialization